inmoose/limma/lmfit.py

```diff
@@ -0,0 +1,433 @@
+    if ne is not None:
+        logging.info(f"Coefficients not estimable: {' '.join(ne)}")
+
+    # Check ndups and spacing. Default to 1.
+    if ndups is None:
+        try:
+            ndups = y.printer.ndups
+        except AttributeError:
+            pass
+    if ndups is None:
+        ndups = 1
+    if spacing is None:
+        try:
+            spacing = y.printer.spacing
+        except AttributeError:
+            pass
+    if spacing is None:
+        spacing = 1
+
+    # Check weights
+    if weights is None:
+        weights = y.weights
+
+    # Check method
+    if method not in ["ls", "robust"]:
+        raise ValueError(f"unknown method: {method}")
+
+    # If duplicates are present, reduce probe-annotation and Amean to correct length
+    if ndups > 1:
+        if y.probes is not None:
+            y.probes = uniquegenelist(y.probes, ndups=ndups, spacing=spacing)
+        if y.Amean is not None:
+            y.Amean = np.nanmean(
+                unwrapdups(y.Amean, ndups=ndups, spacing=spacing), axis=1
+            )
+
+    # Dispatch fitting algorithms
+    if method == "robust":
+        raise NotImplementedError("robust method for lmFit is not implemented yet")
+        # fit = mrlm(
+        #    y.expres, design=design, ndups=ndups, spacing=spacing, weights=weights
+        # )
+    else:
+        if ndups < 2 and block is None:
+            fit = lm_series(
+                y.exprs, design=design, ndups=ndups, spacing=spacing, weights=weights
+            )
+        else:
+            if correlation is None:
+                raise ValueError(
+                    "the correlation must be set, see function duplicateCorrelation"
+                )
+            raise NotImplementedError("gls_series is not implemented yet")
+            # fit = gls_series(
+            #    y.exprs,
+            #    design=design,
+            #    ndups=ndups,
+            #    spacing=spacing,
+            #    block=block,
+            #    correlation=correlation,
+            #    weights=weights,
+            # )
+
+    # Possible warning on missing coefs
+    if fit.coefficients.shape[1] > 1:
+        n = np.sum(np.isnan(fit.coefficients), axis=1)
+        n = np.sum((n > 0) & (n < fit.coefficients.shape[1]))
+        if n > 0:
+            logging.warnings.warn(f"Partial NA coefficients for {n} probes")
+
+    # output
+    fit.genes = y.probes
+    fit.Amean = y.Amean
+    fit.method = method
+    fit.design = design
+    return fit
+
+
+def lm_series(M, design=None, ndups=1, spacing=1, weights=None):
+    """
+    Fit linear model to microarray data by ordinary least squares
+
+    Fit a linear model genewise to expression data from a series of arrays.
+    This function uses ordinary least squares and is a utility function for
+    :func:`lmFit`. Most users should not use this function directly but should
+    use :func:`lmFit` instead.
+
+    The linear model is fit for each gene by calling the function TODO or TODO.
```

Comment:
The TODO functions are not implemented yet, right?

Reply:
this is fixed in #48